internal/allocator/allocator.go

```diff
@@ -245,16 +299,178 @@ func (a *Allocator) Unassign(svc string) {
 	stats.ipv6PoolActive.WithLabelValues(al.pool).Set(float64(len(a.poolIPV6InUse[al.pool])))
 }
 
+// getFreeIPsFromPool determines, with best effort, an ipv4 and an ipv6 available from the provided pool.
+func (a *Allocator) getFreeIPsFromPool(
+	pool *config.Pool,
+	svcKey string,
+	ports []Port,
+	sharingKey,
+	backendKey string,
+) *ipPool {
+	poolIps := newIPPool(pool.Name)
+	for _, cidr := range pool.CIDR {
+		cidrIPFamily := ipfamily.ForCIDR(cidr)
+		if _, ok := poolIps.IPMap[cidrIPFamily]; ok {
+			continue
+		}
+		ip := a.getIPFromCIDR(cidr, pool.AvoidBuggyIPs, svcKey, ports, sharingKey, backendKey)
+		if ip != nil {
+			poolIps.IPMap[cidrIPFamily] = ip
+		}
+	}
+	return poolIps
+}
+
+// findBestPoolForService returns the ipPool corresponding with the most suitable pool for a serviceIPFamily.
+func (a *Allocator) findBestPoolForService(
+	pools []*config.Pool,
+	svcKey string,
+	svc *v1.Service,
+	serviceIPFamily ipfamily.Family,
+	ports []Port,
+	sharingKey, backendKey string,
+) (*ipPool, error) {
+	var primaryIPPoolCandidate, secondaryIPPoolCandidate *ipPool
+	// By default, ipv4 has higher priority.
+	primaryIPFamily := ipfamily.IPv4
+	secondaryIPFamily := ipfamily.IPv6
+	serviceIPFamilyPolicy := v1.IPFamilyPolicySingleStack
+	if svc.Spec.IPFamilyPolicy != nil {
+		serviceIPFamilyPolicy = *(svc.Spec.IPFamilyPolicy)
+	}
+	// If the ipv6 is explicitly required to be prefferable.
+	if len(svc.Spec.IPFamilies) > 0 && svc.Spec.IPFamilies[0] == v1.IPv6Protocol {
+		primaryIPFamily = ipfamily.IPv6
+		secondaryIPFamily = ipfamily.IPv4
+	}
+	for _, pool := range pools {
+		poolIps := a.getFreeIPsFromPool(pool, svcKey, ports, sharingKey, backendKey)
+		// This can happen only in case serviceIPFamily is ipv4 or ipv6.
+		if _, ok := poolIps.IPMap[serviceIPFamily]; ok {
+			return poolIps, nil
+		}
+
+		_, hasPrimaryIP := poolIps.IPMap[primaryIPFamily]
+		_, hasSecondaryIP := poolIps.IPMap[secondaryIPFamily]
+
+		if hasPrimaryIP && hasSecondaryIP {
+			return poolIps, nil
+		}
+
+		// Only continue if policy is preferDualStack and serviceIPFamily is DualStack.
+		if serviceIPFamilyPolicy != v1.IPFamilyPolicyPreferDualStack ||
+			serviceIPFamily != ipfamily.DualStack {
+			continue
+		}
+
+		if hasPrimaryIP && primaryIPPoolCandidate == nil {
+			primaryIPPoolCandidate = poolIps
+		}
+		if hasSecondaryIP && secondaryIPPoolCandidate == nil {
+			secondaryIPPoolCandidate = poolIps
+		}
+	}
+	if primaryIPPoolCandidate != nil {
+		return primaryIPPoolCandidate, nil
+	}
+	if secondaryIPPoolCandidate != nil {
+		return secondaryIPPoolCandidate, nil
+	}
+	return nil, fmt.Errorf("no suitable pool for %s IPFamily", serviceIPFamily)
+}
+
+// allocateBestPool picks the most suitable pool and tries to allocate its ips.
```

Comment:
please invert the order of callee / caller (https://kentcdodds.com/blog/newspaper-code-structure)